keps/sig-node/4210-max-image-gc-age/README.md

```diff
@@ -0,0 +1,373 @@
+However, having the Kubelet be purely reactive means that images will clutter the disk and cause it to fill up. While there aren't reported cases
+where this causes issues, it is inefficient with disk space and can cause the Kubelet to scramble to save disk space when the threshold is met.
+
+An additional approach is to define a way for an admin to request images are cleaned up after they're unused for a certain period of time.
+This would reduce the frequency of the disk usage hitting the level, and provide an admin more flexibility in how garbage collection is defined.
+
+The proposal of this KEP is to add an option to the KubeletConfiguration object that looks like:
+```
+	// ImageMaximumGCAge is the maximum age an image can be unused before it is garbage collected.
+    // The default of this field is 0, which disables it.
+    // +optional
+	ImageMaximumGCAge metav1.Duration
+```
+
+To begin, this option will be set to 0, which will be interpreted as "disabled". In the future, a more reasonable default may be chosen.
+
+### User Stories (Optional)
+
+#### Story 1
+
+- As a cluster admin, I would like my unused images to be garbage collected in a timely manner, and not occupy disk space forever.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- If set incorrectly the ImageMaximumGCAge option could cause unneeded image pulls. For instance, if a Cron job ran
+once a week, but the ImageMaximumGCAge was set to less than a week, that image would get pulled every week, causing needless
+traffic from the registry
+    - Proper documentation on this is the best way to mitigate this risk.
+    - Defining a good default for this value will be similarly tricky.
+
+## Design Details
+
+Add an option to the Kubelet configuration:
+```
+	// ImageMaximumGCAge is the maximum age an image can be unused before it is garbage collected.
+    // The default of this field is 0, which disables it.
+    // +optional
+	ImageMaximumGCAge metav1.Duration
+```
+
+This option will be wired down to the Kubelet's [image manager](https://github.com/kubernetes/kubernetes/blob/d5690f12b69a/pkg/kubelet/images/image_gc_manager.go),
+similarly to the other garbage collection fields.
+
+The Kubelet's image manager already keeps track of the last time an image was used through the `lastUsed` field in the
+[imageRecord](https://github.com/kubernetes/kubernetes/blob/d5690f12b69a/pkg/kubelet/images/image_gc_manager.go#L153) structure.
+So a comparison can be made in the realImageGCManager's function
+[GarbageCollect](https://github.com/kubernetes/kubernetes/blob/d5690f12b69a/pkg/kubelet/images/image_gc_manager.go#L288) to garbage collect
```

Comment:
how often GC will be called?

Reply:
every 5 minutes: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L194

Reply:
can you add a note about it to the doc. So it is clear for the next reader